### ISSUE TITLE
key lookup error in Data.Map

### ISSUE THREAD

**Issue by rzil** [haskell/containers#343]
~~~~
{-
Why does the function "crash" give an error?
Tested on: The Glorious Glasgow Haskell Compilation System, version 8.0.1
-}

module MapCrash where

import Data.Map

g n = fromAscList [("e" ++ show i,("v" ++ show i, "v" ++ show (1 + mod i n))) | i <- [1..n]]

g10 = g 10

{-
When calling the below function I get:
*** Exception: Map.!: given key is not an element in the map
CallStack (from HasCallStack):
  error, called at libraries/containers/Data/Map/Base.hs:489:16 in containers-0.5.7.1:Data.Map.Base
-}
crash = g10 ! ((keys g10) !! 9)
~~~~

**Comment by rzil** [comment]
My mistake. Should use fromList instead of fromAscList.

**Comment by treeowl** [comment]
I wonder if we should use rewrite rules to try to catch these kinds of
mistakes for types with comparison operations around as fast as their
equality operations. We could write a function

fromAscListChecked :: Ord k => [(k,a)] -> Map k a

and then rewrite fromAscList to fromAscListChecked for, say, Int, Int8,
..., Word, Word8, ..., Char, Bool, Integer, Natural, and String. The error
message would warn that the user got lucky and should not rely on the error
being detected.

On Sep 23, 2016 10:02 PM, "Ruben Zilibowitz" <notifications@github.com>
wrote:

Closed #343 <https://github.com/haskell/containers/issues/343>.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/haskell/containers/issues/343#event-801139651>, or mute
the thread

**Comment by rzil** [comment]
Might prevent this from happening to other unsuspecting developers.
